### ISSUE TITLE
Missing method in mono(v3.2.8) for rix0rrr.BeaconLib

### ISSUE THREAD

**Issue by ltem** [OpenRA/OpenRA#13348]
Hello,

I am unable to start skirmish matches on bleed. The moment I click the Skirmish Button, OpenRA crashes. It also happens if I want to start a mission or enter the multiplayer. All mods are affected. The terminal log says something about a 'missing method in mono referenced by rix0rrr.BeaconLib'. 

Exception Log and Terminal Output are attached.

**Edit**: I fixed this issue by switching from the shipped mono version (v3.2.8) to a self-compiled one (v4.8.1). However, it doesn't seem to be the intended way to go.

Exception log:
```
Dune 2000 Mod at Version {DEV_VERSION}
on map c878e27c993e5882823d5bd51ce05c7ff6aed770 (D2k Shellmap by Westwood Studios, MustaphaTR).
Date: 2017-05-24 18:25:50Z
Operating System: Linux (Unix 3.16.0.4)
Runtime Version: Mono 3.2.8 (Debian 3.2.8+dfsg-10) CLR 4.0.30319.17020
Exception of type `System.Reflection.TargetInvocationException`: Exception has been thrown by the target of an invocation.
Inner
  Exception of type `System.TypeInitializationException`: An exception was thrown by the type initializer for OpenRA.Mods.Common.Server.MasterServerPinger
  Inner
    Exception of type `System.MissingMethodException`: Method not found: 'System.Net.Sockets.UdpClient.AllowNatTraversal'.
      at OpenRA.Mods.Common.Server.MasterServerPinger..cctor () [0x00000] in <filename unknown>:0
    at (wrapper managed-to-native) System.Reflection.MonoCMethod:InternalInvoke (System.Reflection.MonoCMethod,object,object[],System.Exception&)
  at System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters) [0x00000] in <filename unknown>:0
  at System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters) [0x00000] in <filename unknown>:0
  at System.Reflection.MonoCMethod.DoInvoke (System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00000] in <filename unknown>:0
  at System.Reflection.MonoCMethod.Invoke (BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00000] in <filename unknown>:0
  at System.Reflection.ConstructorInfo.Invoke (System.Object[] parameters) [0x00000] in <filename unknown>:0
  at OpenRA.ObjectCreator.CreateBasic (System.Type type) [0x00000] in <filename unknown>:0
  at OpenRA.ObjectCreator.CreateObject[ServerTrait] (System.String className, System.Collections.Generic.Dictionary`2 args) [0x00000] in <filename unknown>:0
  at OpenRA.ObjectCreator.CreateObject[ServerTrait] (System.String className) [0x00000] in <filename unknown>:0
  at OpenRA.Server.Server..ctor (System.Net.IPEndPoint endpoint, OpenRA.ServerSettings settings, OpenRA.ModData modData, Boolean dedicated) [0x00000] in <filename unknown>:0
  at OpenRA.Game.CreateLocalServer (System.String map) [0x00000] in <filename unknown>:0
  at OpenRA.Mods.Common.Widgets.Logic.MainMenuLogic.StartSkirmishGame () [0x00000] in <filename unknown>:0
  at OpenRA.Mods.Common.Widgets.ButtonWidget.<ButtonWidget>m__B (MouseInput _) [0x00000] in <filename unknown>:0
  at OpenRA.Mods.Common.Widgets.ButtonWidget.HandleMouseInput (MouseInput mi) [0x00000] in <filename unknown>:0
  at OpenRA.Widgets.Widget.HandleMouseInputOuter (MouseInput mi) [0x00000] in <filename unknown>:0
  at OpenRA.Widgets.Ui.HandleInput (MouseInput mi) [0x00000] in <filename unknown>:0
  at OpenRA.DefaultInputHandler+<OnMouseInput>c__AnonStorey2.<>m__0 () [0x00000] in <filename unknown>:0
  at OpenRA.Sync.CheckSyncUnchanged[Boolean] (OpenRA.World world, System.Func`1 fn) [0x00000] in <filename unknown>:0
  at OpenRA.DefaultInputHandler.OnMouseInput (MouseInput input) [0x00000] in <filename unknown>:0
  at OpenRA.Platforms.Default.Sdl2Input.PumpInput (OpenRA.Platforms.Default.Sdl2GraphicsDevice device, IInputHandler inputHandler) [0x00000] in <filename unknown>:0
  at OpenRA.Platforms.Default.Sdl2GraphicsDevice.PumpInput (IInputHandler inputHandler) [0x00000] in <filename unknown>:0
  at OpenRA.Renderer.EndFrame (IInputHandler inputHandler) [0x00000] in <filename unknown>:0
  at OpenRA.Game.RenderTick () [0x00000] in <filename unknown>:0
  at OpenRA.Game.Loop () [0x00000] in <filename unknown>:0
  at OpenRA.Game.Run () [0x00000] in <filename unknown>:0
  at OpenRA.Program.Run (System.String[] args) [0x00000] in <filename unknown>:0
  at OpenRA.Program.Main (System.String[] args) [0x00000] in <filename unknown>:0
```

Terminal output:
```
Platform is Linux
Using SDL 2 with OpenGL renderer
Desktop resolution: 1920x1080
Using resolution: 1400x1024
Using window scale 1.00
OpenGL version: 4.5.13399 Compatibility Profile Context 15.201.1151
Using default sound device
Internal mods:
        all: All mods ({DEV_VERSION})
        cnc: Tiberian Dawn ({DEV_VERSION})
        d2k: Dune 2000 ({DEV_VERSION})
        modcontent: Mod Content Manager ({DEV_VERSION})
        ra: Red Alert ({DEV_VERSION})
        ts: Tiberian Sun ({DEV_VERSION})
External mods:
        cnc-{DEV_VERSION}: Tiberian Dawn ({DEV_VERSION})
        d2k-git-65708b4: Dune 2000 (git-65708b4)
        d2k-{DEV_VERSION}: Dune 2000 ({DEV_VERSION})
        ra-{DEV_VERSION}: Red Alert ({DEV_VERSION})
        ts-{DEV_VERSION}: Tiberian Sun ({DEV_VERSION})
Loading mod: d2k
Missing method System.Net.Sockets.UdpClient::AllowNatTraversal(bool) in assembly /usr/lib/mono/gac/System/4.0.0.0__b77a5c561934e089/System.dll, referenced in assembly /home/user/workspace/OpenRA-Project/OpenRA/rix0rrr.BeaconLib.dll
Exception of type `System.Reflection.TargetInvocationException`: Exception has been thrown by the target of an invocation.
Inner
  Exception of type `System.TypeInitializationException`: An exception was thrown by the type initializer for OpenRA.Mods.Common.Server.MasterServerPinger
  Inner
    Exception of type `System.MissingMethodException`: Method not found: 'System.Net.Sockets.UdpClient.AllowNatTraversal'.
      at OpenRA.Mods.Common.Server.MasterServerPinger..cctor () [0x00000] in <filename unknown>:0 
    at (wrapper managed-to-native) System.Reflection.MonoCMethod:InternalInvoke (System.Reflection.MonoCMethod,object,object[],System.Exception&)
  at System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters) [0x00000] in <filename unknown>:0 
  at System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters) [0x00000] in <filename unknown>:0 
  at System.Reflection.MonoCMethod.DoInvoke (System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00000] in <filename unknown>:0 
  at System.Reflection.MonoCMethod.Invoke (BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00000] in <filename unknown>:0 
  at System.Reflection.ConstructorInfo.Invoke (System.Object[] parameters) [0x00000] in <filename unknown>:0 
  at OpenRA.ObjectCreator.CreateBasic (System.Type type) [0x00000] in <filename unknown>:0 
  at OpenRA.ObjectCreator.CreateObject[ServerTrait] (System.String className, System.Collections.Generic.Dictionary`2 args) [0x00000] in <filename unknown>:0 
  at OpenRA.ObjectCreator.CreateObject[ServerTrait] (System.String className) [0x00000] in <filename unknown>:0 
  at OpenRA.Server.Server..ctor (System.Net.IPEndPoint endpoint, OpenRA.ServerSettings settings, OpenRA.ModData modData, Boolean dedicated) [0x00000] in <filename unknown>:0 
  at OpenRA.Game.CreateLocalServer (System.String map) [0x00000] in <filename unknown>:0 
  at OpenRA.Mods.Common.Widgets.Logic.MainMenuLogic.StartSkirmishGame () [0x00000] in <filename unknown>:0 
  at OpenRA.Mods.Common.Widgets.ButtonWidget.<ButtonWidget>m__B (MouseInput _) [0x00000] in <filename unknown>:0 
  at OpenRA.Mods.Common.Widgets.ButtonWidget.HandleMouseInput (MouseInput mi) [0x00000] in <filename unknown>:0 
  at OpenRA.Widgets.Widget.HandleMouseInputOuter (MouseInput mi) [0x00000] in <filename unknown>:0 
  at OpenRA.Widgets.Ui.HandleInput (MouseInput mi) [0x00000] in <filename unknown>:0 
  at OpenRA.DefaultInputHandler+<OnMouseInput>c__AnonStorey2.<>m__0 () [0x00000] in <filename unknown>:0 
  at OpenRA.Sync.CheckSyncUnchanged[Boolean] (OpenRA.World world, System.Func`1 fn) [0x00000] in <filename unknown>:0 
  at OpenRA.DefaultInputHandler.OnMouseInput (MouseInput input) [0x00000] in <filename unknown>:0 
  at OpenRA.Platforms.Default.Sdl2Input.PumpInput (OpenRA.Platforms.Default.Sdl2GraphicsDevice device, IInputHandler inputHandler) [0x00000] in <filename unknown>:0 
  at OpenRA.Platforms.Default.Sdl2GraphicsDevice.PumpInput (IInputHandler inputHandler) [0x00000] in <filename unknown>:0 
  at OpenRA.Renderer.EndFrame (IInputHandler inputHandler) [0x00000] in <filename unknown>:0 
  at OpenRA.Game.RenderTick () [0x00000] in <filename unknown>:0 
  at OpenRA.Game.Loop () [0x00000] in <filename unknown>:0 
  at OpenRA.Game.Run () [0x00000] in <filename unknown>:0 
  at OpenRA.Program.Run (System.String[] args) [0x00000] in <filename unknown>:0 
  at OpenRA.Program.Main (System.String[] args) [0x00000] in <filename unknown>:0 

AL lib: (WW) FreeContext: (0x17f4940) Deleting 32 Source(s)
AL lib: (WW) FreeDevice: (0x180c1a0) Deleting 2 Buffer(s)
Gtk-Message: GtkDialog mapped without a transient parent. This is discouraged.
```

**Comment by pchote** [comment]
We should catch these errors and disable the LAN game advertising instead of crashing.

**Comment by rob-v** [comment]
I think it can still work without AllowNatTraversal. I can't verify it in advance, but preparing PR to resolve this issue in rix0rrr.BeaconLib. We could also update OpenRA to ignore other possible errors and disable this feature.

**Comment by rob-v** [comment]
Fixed by #13351

There is another PR to update BeaconLib to support also frameworks with missing AllowNatTraversal, which I assume is not mandatory for this feature.
